_posts/2017-11-27-Babel-7-and-TypeScript.markdown

```diff
@@ -0,0 +1,261 @@
+
+```json
+"resolutions": {
+  "babel-core": "^7.0.0-bridge.0"
+},
+```
+
+With this in place many of our errors disappeared and packages like `jest` now worked like a charm.
+
+### 2) Missing ES2015 Features
+
+Another error we faced early on surrounded language features that worked with Babel _or_ TypeScript, but not with Babel _and_ TypeScript. For example, take an existing Babel project that points to `index.js` as an entrypoint, configure it to support TypeScript via Babel 7, and then run it:
+
+```json
+"scripts": {
+  "start": "babel-node index.js"
+}
+```
+```js
+// index.js
+require('@babel/register', {
+  extensions: ['.js', '.jsx', '.ts', '.tsx']
+})
+require('app/server.ts')
+```
+```typescript
+// app/server.ts
+console.log('hi!')
+```
+
+Running
+
+```sh
+yarn start
+$ babel-node index.js
+
+hi!
+✨  Done in 1.88s.
+```
+
+Everything seems to be working; our `.js` entrypoint is configured to support `.ts` extensions and we kick off the boot process.
+
+Let's now try to import a file from within `app/server.ts`:
+
+```typescript
+import path from 'path'
+console.log(`Hello ${path.resolve(process.cwd())}!`)
+```
+
+```sh
+yarn start
+$ yarn run v1.3.2
+$ babel-node index.js
+sites/src/index.tsx:1
+(function (exports, require, module, __filename, __dirname) { import path from 'path'
+                                                              ^^^^^^
+
+SyntaxError: Unexpected token import
+```
+
+Maybe my `tsconfig.json` file is misconfigured?
+
+```json
+{
+  "compilerOptions": {
+    "module": "es2015"
+  }
+}
+```
+
+Nope, all good. How about my `.babelrc`?
+
+```json
+{
+  "presets": [
+    ["@babel/env", {
+      "targets": {
+        "browsers": ["last 2 versions"]
+      }
+    }],
+    "@babel/stage-3",
+    "@babel/react",
+    "@babel/typescript"
+  ]
+}
+```
+
+We're using [`@babel/preset-env`](https://github.com/babel/babel/tree/master/packages/babel-preset-env) which handles selecting the JS features we need, so thats not it. And anyways, doesn't TypeScript support `ES2015` modules right out of the box?
+
+Continuing, how about specifying the extension list directly in `package.json`:
+
+```json
+"start": "babel-node --extensions '.ts,.tsx' index.js"
+```
+
+Still no go :(
+
+Last try: Create a new entrypoint file that uses a `.ts` extension and then use _that_ to boot the rest of the app:
+
+```json
+"start": "babel-node --extensions '.ts,.tsx' index.ts"
+```
+```typescript
+// index.ts
+import './app/server'
+```
+```sh
+yarn start
+$ yarn run v1.3.2
+$ babel-node index.js
+Hello /sites!
+```
+
+Once this change was in place, we could ditch `@babel/register` and instead rely on the `--extensions` configuration from `package.json`.
```

Comment:
A few notes:

* Can you link to some babel tickets on the topic of babel not always loading `ts` files when you’d might expect that?
* While it is good to understand this issue when running into it, the opinion of the babel team, and mine, is still that [runtime compilation should be avoided](https://hashnode.com/ama/with-babel-team-cj7awmk5e00ij54wu6onnyc5w#cj7be3u8e00t554wu7nti3iiu), so getting rid of it now is a win-win, which may make the reader feel less like they’re doing work to workaround babel ‘limitations’.

Reply:
> While it is good to understand this issue when running into it, the opinion of the babel team, and mine, is still that runtime compilation should be avoided, so getting rid of it now is a win-win, which may make the reader feel less like they’re doing work to work around babel ‘limitations’.

I would argue that `babel-node` is a critical part of a modern development flow, and getting that set up leads to vastly improved iteration times since a project can, from the start, be built out in a standard language across client / server with shared features and zero conflicts. I would be open to including a comment on transpiling code for production, but I think that individual circumstance should dictate that decision rather than walking the reader through more complexity for an end result that's beyond the scope of this post.

Reply:
Oh absolutely, my bad for not being more specific about what dependency I was talking. `babel-node` ftw in development, just not `babel-register` because it hardcodes the dependency on your app code.

Reply:
But you’re absolutely right in not needing to dive into production details 👍